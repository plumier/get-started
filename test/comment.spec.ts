import "@plumier/typeorm"

import dotenv from "dotenv"
import { join } from "path"
import supertest from "supertest"
import { getConnection } from "typeorm"

import createApp from "../src/app"
import { adminToken, createUser, ignoreProps, userToken } from "./shared"


async function createTodo(app: any, userToken: string) {
    const { body } = await await supertest(app.callback())
        .post("/api/v1/todos")
        .send({ message: "lorem ipsum dolor sit amet" })
        .set("Authorization", `Bearer ${userToken}`)
    return { id: body.id }
}

dotenv.config({ path: join(__dirname, ".env-test") })

afterEach(async () => {
    const con = getConnection()
    if (con.isConnected)
        await con.close()
})

describe("Comment Spec", () => {
    it("Should not allow non login user to comment", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const todo = await createTodo(app, user.token)
        await supertest(app.callback())
            .post(`/api/v1/todos/${todo.id}/comments`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .expect(403)
    })
    it("Should automatically add comment owner with current login user", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const todo = await createTodo(app, user.token)
        const { body: comment } = await supertest(app.callback())
            .post(`/api/v1/todos/${todo.id}/comments`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/v1/todos/${todo.id}/comments/${comment.id}`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        expect(body).toMatchSnapshot({ ...ignoreProps, user: ignoreProps })
    })
    it("Should only the owner of the comment or Admin can modify", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const todo = await createTodo(app, user.token)
        const { body: comment } = await supertest(app.callback())
            .post(`/api/v1/todos/${todo.id}/comments`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/todos/${todo.id}/comments/${comment.id}`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/todos/${todo.id}/comments/${comment.id}`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${adminToken}`)
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/todos/${todo.id}/comments/${comment.id}`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${userToken}`)
            .expect(401)
    })
    it("Should only the owner of the comment or Admin can delete", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const todo = await createTodo(app, user.token)
        const { body: comment } = await supertest(app.callback())
            .post(`/api/v1/todos/${todo.id}/comments`)
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .delete(`/api/v1/todos/${todo.id}/comments/${comment.id}`)
            .set("Authorization", `Bearer ${userToken}`)
            .expect(401)
        await supertest(app.callback())
            .delete(`/api/v1/todos/${todo.id}/comments/${comment.id}`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
    })
})