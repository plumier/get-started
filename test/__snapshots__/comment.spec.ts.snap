// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Spec Should automatically add comment owner with current login user 1`] = `
Object {
  "createdAt": Any<String>,
  "deleted": false,
  "id": 1,
  "message": "lorem ipsum dolor sit amet",
  "updatedAt": Any<String>,
  "user": Object {
    "createdAt": Any<String>,
    "deleted": false,
    "email": "john.doe@gmail.com",
    "id": 1,
    "name": "John Doe",
    "role": "User",
    "updatedAt": Any<String>,
  },
}
`;
